### PR TITLE
ARROW-12827: [C++] Improve error message for dataset discovery failure

### PR DESCRIPTION
**Pull request by lidavidm**
This adds a bit more context to the error messages, though maybe this is a bit wordy?

```
>>> ds.dataset('dataset4', format="ipc")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lidavidm/Code/upstream/arrow-12827/python/pyarrow/dataset.py", line 655, in dataset
    return _filesystem_dataset(source, **kwargs)
  File "/home/lidavidm/Code/upstream/arrow-12827/python/pyarrow/dataset.py", line 410, in _filesystem_dataset
    return factory.finish(schema)
  File "pyarrow/_dataset.pyx", line 2262, in pyarrow._dataset.DatasetFactory.finish
    return Dataset.wrap(GetResultValue(result))
  File "pyarrow/error.pxi", line 141, in pyarrow.lib.pyarrow_internal_check_status
    return check_status(status)
  File "pyarrow/error.pxi", line 97, in pyarrow.lib.check_status
    raise ArrowInvalid(message)
pyarrow.lib.ArrowInvalid: Error creating dataset. Could not read schema from 'dataset4/foo.parquet': Could not open IPC input source 'dataset4/foo.parquet': File is too small: 9. Is this a 'ipc' file?
>>> ds.dataset('dataset5', format="parquet")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lidavidm/Code/upstream/arrow-12827/python/pyarrow/dataset.py", line 655, in dataset
    return _filesystem_dataset(source, **kwargs)
  File "/home/lidavidm/Code/upstream/arrow-12827/python/pyarrow/dataset.py", line 410, in _filesystem_dataset
    return factory.finish(schema)
  File "pyarrow/_dataset.pyx", line 2262, in pyarrow._dataset.DatasetFactory.finish
    return Dataset.wrap(GetResultValue(result))
  File "pyarrow/error.pxi", line 141, in pyarrow.lib.pyarrow_internal_check_status
    return check_status(status)
  File "pyarrow/error.pxi", line 112, in pyarrow.lib.check_status
    raise IOError(message)
OSError: Error creating dataset. Could not read schema from 'dataset5/foo.parquet': Could not open Parquet input source 'dataset5/foo.parquet': Invalid: Parquet magic bytes not found in footer. Either the file is corrupted or this is not a parquet file.. Is this a 'parquet' file?
```

### CODE REVIEW

File: cpp/src/arrow/dataset/file_parquet.cc [hunk 1]
```diff
@@ -311,7 +311,7 @@ Result<std::unique_ptr<parquet::arrow::FileReader>> ParquetFileFormat::GetReader
   try {
     reader = parquet::ParquetFileReader::Open(std::move(input), std::move(properties));
   } catch (const ::parquet::ParquetException& e) {
-    return Status::IOError("Could not open parquet input source '", source.path(),
+    return Status::IOError("Could not open Parquet input source '", source.path(),
```
Comment (westonpace) [inline cpp/src/arrow/dataset/file_parquet.cc:314]: I haven't really checked but are we consistent with capitalizing parquet in this way?  It's just a nit but I think I've used lowercase for file types.  E.g. an xml file or a json file.  Although I suppose it's to match the file extension in those cases.

**Review by westonpace — APPROVED** [review]
It's does not sound too wordy to me.  If anything we could maybe drop the "Is this a XYZ file?".  Everything else is adding information the user could not easily get otherwise.

File: cpp/src/arrow/dataset/file_parquet.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/dataset/file_parquet.cc:314]: I'm not too sure either but I can just revert this, there are also possibly other nits like whether we want to report IPC as 'IPC' or 'Feather'.

I'll also drop the 'Is this a XYZ file', I think it's clear enough with the rest of the context now.

**Comment by jorisvandenbossche** [comment]
I think the "Is this a XYZ file?" is actually quite useful, and not too verbose. Because it's something easy to get when reading a non-parquet file and you forget to specify the format (the default is "parquet", and not to infer it from the file, as users might expect)

This relates to a PR @thisisnic did for improving this error message on the R side -> https://github.com/apache/arrow/pull/10326 (this PR might cover the custom handling you added in R?)

**Comment by jorisvandenbossche** [comment]
> I think the "Is this a XYZ file?" is actually quite useful, and not too verbose

Of course the "Could not open Parquet input source" part also already gives a hint for that

**Comment by lidavidm** [comment]
This is even more wordy, but perhaps `If reading a different format than 'Parquet', pass the intended format to the dataset/factory`?

**Comment by pitrou** [comment]
I think the "Is this a XYZ file?" conveys the information quite clearly (and invites the user to check).

**Comment by lidavidm** [comment]
Alright, I added back the 'Is this a XYZ file' message.

**Comment by lidavidm** [comment]
Is everyone happy with the error message here? :slightly_smiling_face:
